Superb guide: Handle 404s when getting users/teams, proxy the teams page

GregWeil opened the pull request:
The api is going to start giving 404s when teams don't exist rather than returning `null`.

I also did some cleanup around the proxy so we can use Pirijian's new teams page. Proxied sites can exist at their root url now, but they'll have to be careful not to use any base relative urls.

judeallred commented:
Prior to this change, only Ghost blogs has the "url must end in a '/'" constraint.  This is because Ghost loves relative urls so we need that slash to guard us,  however pages that aren't in ghost, such as https://glitch.com/forplatforms and https://glitch.com/teams  were permitted without the trailing slash.

With this change, the trailing slash is now mandatory on all the proxied routes, although not on glitch.com sites themselves.

Not sure if intentional or regression?

judeallred commented:
appropos of ^^;   the redirects (to the ghost sites) should hall have trailing slashes on them otherwise we get the double redirect effect

GregWeil commented:
Fixed a regression with /faq. (Server redirecting broke some of the hash url redirects, such as glitch.com/faq#gfp )
Also improved console logging when the paths don't match (it now prints the requested url, not where it got proxied to)
Also fixed that double redirect issue.

The / is required for the new /teams route, but I guess I could make it specific to that as well.

GregWeil commented on src/presenters/pages/join-team.jsx:
```diff
@@ -19,7 +19,9 @@ class JoinTeamPageBase extends React.Component {
     } catch (error) {
-      Raven.captureException(error);
+      if (error && !(error.response && error.response.status === 404)) {
+        Raven.captureException(error);
```
It happens automatically for any uncaught errors.

judeallred commented on server/proxy.js:
```diff
@@ -39,27 +25,40 @@ module.exports = function(app) {
+  function proxyGhost(route, glitchTarget, pathOnTarget="") {
     // Proxy all the requests to /{route}/ over to glitchTarget
-    proxyGlitch(route, glitchTarget, pathOnTarget);
+    proxyGlitch(route, glitchTarget, urlJoin("/", pathOnTarget, route));
```
url-join doesn't normalize;  it'll put a slash on the end if the final string in the list ends with a slash

pketh commented on server/redirects.js:
```diff
@@ -1,7 +1,7 @@
-  redirect(app, '/forteams*', '/teams');
+  redirect(app, '/partners*', '/teams/');
+  redirect(app, '/foryourapi*', '/teams/');
+  redirect(app, '/forteams*', '/teams/');
```
will `glitch.com/teams` work?

GregWeil replied:
It would but we redirect /teams to /teams/ and it's faster to just skip the second redirect.

pketh replied:
the second redirect is conditional though, so if `/teams/` is used it'll just redirect once, 

but if `/teams` is used (as a fallback, or mistakenly), then the second redirect happens (and the speed penalty is paid only then)?

judeallred commented on server/redirects.js:
```diff
@@ -1,7 +1,7 @@
+  redirect(app, '/partners*', '/teams/');
+  redirect(app, '/foryourapi*', '/teams/');
+  redirect(app, '/forteams*', '/teams/');
   redirect(app, '/website-starter-kit*', '/featured/website-starter-kit');  
```
these are all double-redirects too

GregWeil commented:
@garethx All of the proxied sites should work as is. /forplatforms is still loaded from about.glitch.me, and /faq does some special redirecting based on hash urls (`/faq#gfp` is the only one I know)
